src/legacy/huffman.rs

```diff
@@ -0,0 +1,215 @@
+use std::io::{self, Error};
+
+use crate::legacy::bitstream::reverse_lsb;
+
+use super::bitstream::lsb;
+
+#[derive(Default, Clone, Copy)]
+pub struct TableEntry {
+    /// Wide enough to fit the max symbol nbr.
+    pub sym: u16,
+    /// 0 means no symbol.
+    pub len: u8,
+}
+
+/// Deflate uses max 288 symbols.
+const MAX_HUFFMAN_SYMBOLS: usize = 288;
+/// Implode uses max 16-bit codewords.
+const MAX_HUFFMAN_BITS: usize = 16;
+/// Seems a good trade-off.
+const HUFFMAN_LOOKUP_TABLE_BITS: u8 = 8;
+
+pub struct HuffmanDecoder {
+    /// Lookup table for fast decoding of short codewords.
+    pub table: [TableEntry; 1 << HUFFMAN_LOOKUP_TABLE_BITS],
+    /// "Sentinel bits" value for each codeword length.
+    pub sentinel_bits: [u32; MAX_HUFFMAN_BITS],
+    /// First symbol index minus first codeword mod 2**16 for each length.
+    pub offset_first_sym_idx: [u16; MAX_HUFFMAN_BITS],
+    /// Map from symbol index to symbol.
+    pub syms: [u16; MAX_HUFFMAN_SYMBOLS],
+    // num_syms:usize
+}
+
+impl Default for HuffmanDecoder {
+    fn default() -> Self {
+        let syms = [0; MAX_HUFFMAN_SYMBOLS];
+        let table = [TableEntry::default(); 1 << HUFFMAN_LOOKUP_TABLE_BITS];
+        Self {
+            table,
+            sentinel_bits: Default::default(),
+            offset_first_sym_idx: Default::default(),
+            syms,
+        }
+    }
+}
+
+/// Initialize huffman decoder d for a code defined by the n codeword lengths.
+/// Returns false if the codeword lengths do not correspond to a valid prefix
+/// code.
+impl HuffmanDecoder {
+    pub fn init(&mut self, lengths: &[u8], n: usize) -> std::io::Result<()> {
+        let mut count = [0; MAX_HUFFMAN_BITS];
+        let mut code = [0; MAX_HUFFMAN_BITS];
+        let mut sym_idx: [u16; 16] = [0; MAX_HUFFMAN_BITS];
+        // Zero-initialize the lookup table.
+        for t in &mut self.table {
+            t.len = 0;
+        }
+
+        // Count the number of codewords of each length.
+        for sym in 0..n {
+            let len = lengths[sym] as usize;
+            // Ignore zero-length codewords.
+            if len == 0 {
+                continue;
+            }
+            debug_assert!(len < MAX_HUFFMAN_BITS);
```

Comment:
I don't think this will hold. Implode uses codewords up to 16 bits long.

Reply:
kk I remove the change - looked like it's safe to go just to MAX_HUFFMAN_BITS but wasn't sure about that.